packages/ts-autocomplete/src/index.ts

```diff
@@ -8,66 +9,145 @@ const debugError = createDebug('ts-autocomplete:error');
   const options: ts.CompilerOptions = {
     target: ts.ScriptTarget.ES2022,
     allowJs: true,
     moduleResolution: ts.ModuleResolutionKind.NodeNext,
-    types: ['node'],
-    lib: ['es2019'],
-    //typeRoots: [],
+    types: [],
+    lib: ['es2023'],
     allowImportingTsExtensions: true,
   };
 
-  const updateCode = (newDef: Record<TypeFilename, string>): void => {
+  const updateCode = (newDef: Record<TypeFilename, string | boolean>): void => {
     for (const [key, value] of Object.entries(newDef)) {
       codeHolder[key] = value;
       versions[key] = (versions[key] ?? 0) + 1;
     }
   };
 
-  const listFiles = () => {
+  const listFiles = (): string[] => {
     return Object.keys(codeHolder);
   };
 
+  const encounteredPaths: EncounteredPaths = {
+    getScriptSnapshot: [],
+    fileExists: [],
+    readFile: [],
+  };
+  const listEncounteredPaths = (): EncounteredPaths => {
+    encounteredPaths.getScriptSnapshot.sort();
+    encounteredPaths.fileExists.sort();
+    encounteredPaths.readFile.sort();
+    return encounteredPaths;
+  };
+
   const servicesHost: ts.LanguageServiceHost = {
     getScriptFileNames: () => {
       return Object.keys(codeHolder);
     },
     getScriptVersion: (fileName) => {
+      fileName = relativeNodePath(fileName);
       return (versions[fileName] ?? 1).toString();
     },
     getScriptSnapshot: (fileName) => {
+      fileName = relativeNodePath(fileName);
       if (fileName in codeHolder) {
-        return ts.ScriptSnapshot.fromString(codeHolder[fileName]);
+        return ts.ScriptSnapshot.fromString(codeHolder[fileName].toString());
       }
 
-      return ts.ScriptSnapshot.fromString(ts.sys.readFile(fileName) || '');
+      if (process.env.CI) {
+        const result = ts.ScriptSnapshot.fromString(
+          // NOTE: some files do not exist. A good example is "typescript/lib/es2023.ts"
+          ts.sys.readFile(fileName) || '',
+        );
+
+        encounteredPaths.getScriptSnapshot.push(relativeNodePath(fileName));
+        return result;
+      }
     },
     getCurrentDirectory: () => process.cwd(),
     getCompilationSettings: () => options,
-    getDefaultLibFileName: (options) => ts.getDefaultLibFilePath(options),
+    getDefaultLibFileName: (options) => {
+      return ts.getDefaultLibFilePath(options);
+    },
     fileExists: (fileName) => {
-      return fileName in codeHolder || ts.sys.fileExists(fileName);
+      fileName = relativeNodePath(fileName);
+      if (fileName in codeHolder) {
+        return true;
+      }
+
+      if (process.env.CI) {
+        const result = ts.sys.fileExists(fileName);
+        if (result) {
+          encounteredPaths.fileExists.push(relativeNodePath(fileName));
+        }
+        return result;
+      }
+
+      return false;
     },
     readFile: (fileName) => {
+      fileName = relativeNodePath(fileName);
       if (fileName in codeHolder) {
-        return codeHolder[fileName];
+        return codeHolder[fileName].toString();
```

Comment:
I'm guessing this is because we now also have these boolean values there for exists case? Would it make sense to throw if you're trying to readFile that we didn't read actually or is it an expected thing to return stringified boolean here? (I'm maybe just not understanding this part fully)